### ISSUE TITLE
[SPIR-V] Compiler does not complain about writing to read only StructuredBuffer<> resources (DXIL backend does)

### ISSUE THREAD

**Issue by Dredhog** [microsoft/DirectXShaderCompiler#3464]
### Title
[SPIR-V] Compiler does not complain about writing to read only StructuredBuffer<> resources (DXIL backend does)

### Functional impact
The SPIR-V backend does not produce a compiler error when `StructuredBuffer` which is a read only type is written into. 

### Minimal repro steps
1. Compile the following shader with `dxc.exe -T ps_6_0 -E PSMain -spirv issue_shader.txt`:
```
struct MyStruct {
    float2 uv;
};

StructuredBuffer<MyStruct> MySB;

float4 PSMain (float2 uv : TEXCOORD0) : SV_Target
{
    MyStruct myStruct;
    myStruct.uv = uv;
    
    MySB[int(uv.x)] = myStruct;
    return float4(1, 1, 1, 1);
};
```
### Expected result
The SPIR-V backend should produce some user readable error notifying about the invalid write operations. e.g. the DXIL backend fails compilation and produces the following validation error:
```
warning: DXIL.dll not found.  Resulting DXIL will not be signed for use in release environments.

error: validation errors
Function: PSMain: error: store should be on uav resource. Use /Zi for source location.
note: at 'call void @dx.op.bufferStore.f32(i32 69, %dx.types.Handle %1, i32 %4, i32 0, float %2, float %3, float undef, float undef, i8 3)' in block '#0' of function 'PSMain'.
Validation failed.
```
Though this is better than no error it is still suboptimal as the error does not make it clear where the issue is in the source.

### Actual result
Shader compiles without any errors.

### Further technical details
Ran this on [shader playground](http://shader-playground.timjones.io/4867b54b850cd6bfc76be207d8541de0) with DXC from 2020 12 16

**Comment by jaebaek** [comment]
@Dredhog Thank you for reporting the issue.

**Comment by jaebaek** [comment]
I tested it with the current tot 9972b664a34e98e99906a1c3e851be64f3cbb926 build and confirmed it is fixed.

```
a.hlsl:6:20: error: cannot assign to return value because function 'operator[]<const unsigned int &>' returns a const value
    MyBuffer[id.x] = id.y;
    ~~~~~~~~~~~~~~ ^
note: function 'operator[]<const unsigned int &>' which returns const-qualified type 'const unsigned int &' declared here
```

The new error message clearly pinpoints the code location causing the error.
